File: app/src/main/java/com/starsep/sokoban/GameActivity.java
package com.starsep.sokoban;

import android.app.Activity;
import android.content.Context;
import android.os.Bundle;
import android.view.View;
import android.widget.TextView;

import com.starsep.sokoban.database.DatabaseManager;
import com.starsep.sokoban.gamelogic.GameModel;
import com.starsep.sokoban.gamelogic.Gameplay;
import com.starsep.sokoban.gamelogic.HighScore;

import java.util.Timer;
import java.util.TimerTask;

public class GameActivity extends Activity implements GameController {
	private GameView gameView;
	private GameModel gameModel;
	private TextView statusTextView;
	private Timer timer;

	protected void onPause() {
		super.onPause();
		onGamePause();
	}

	@Override
	protected void onStart() {
		super.onStart();
		onGameStart();
	}

	@Override
	protected void onCreate(Bundle savedInstanceState) {
		super.onCreate(savedInstanceState);
		boolean newGame = false;
		Bundle extras = getIntent().getExtras();
		if (extras != null) {
			 newGame = extras.getBoolean("New");
			getIntent().putExtra("New", false);
		}

		setContentView(R.layout.activity_game);

		gameView = (GameView) findViewById(R.id.gameView);
		statusTextView = (TextView) findViewById(R.id.statusTextView);
		Gameplay gameplay;
		if (newGame) {
			gameplay = new Gameplay(this);
		} else {
			gameplay = DatabaseManager.instance(this).getCurrentGame(this);
			gameplay.setGameController(this);
		}
		gameplay.setViewListener(gameView);
		gameModel = gameplay;

		gameView.setGameController(this);
		gameView.setGameModel(gameModel);
		gameView.setOnTouchListener(new OnSwipeTouchListener(this) {
			public void onSwipeTop() {
				gameModel.moveUp();
			}

			public void onSwipeRight() {
				gameModel.moveRight();
			}

			public void onSwipeLeft() {
				gameModel.moveLeft();
			}

			public void onSwipeBottom() {
				gameModel.moveDown();
			}
		});
	}


	public void resetButtonClicked(View view) {
		gameModel.repeatLevel();
	}

	public void undoButtonClicked(View view) { gameModel.undoMove(); }

	public void settingsButtonClicked(View view) {

	}

	public void onStatsChanged() {
		int levelNumber = gameModel.levelNumber();
		HighScore highScore = gameModel.stats();
		int minutes = highScore.time / 60;
		int seconds = highScore.time % 60;
		int moves = highScore.moves;
		int pushes = highScore.pushes;
		statusTextView.setText(String.format(getResources().getString(R.string.level_status),
				levelNumber, minutes, seconds, moves, pushes));
	}

	@Override
	public void onGamePause() {
		if (timer != null) {
			timer.cancel();
			timer = null;
		}
	}

	@Override
	public void onGameStart() {
		onGamePause();
		timer = new Timer();
		timer.schedule(new TimerTask() {
			@Override
			public void run() {
				runOnUiThread(() -> gameModel.onSecondElapsed());
			}
		}, 0, 1000);
	}

	@Override
	public void onSaveGame(Gameplay game) {
		DatabaseManager.instance(this).setCurrentGame(game);
	}

	@Override
	public void onNewGame() {
		onGameStart();
	}

	@Override
	public boolean editMode() {
		return gameView.isInEditMode();
	}

	@Override
	public Context getContext() {
		return super.getBaseContext();
	}
}

File: app/src/main/java/com/starsep/sokoban/release/activity/GameActivity.kt
package com.starsep.sokoban.release.activity

import android.arch.lifecycle.Observer
import android.arch.lifecycle.ViewModelProviders
import android.content.Intent
import android.os.Bundle
import com.starsep.sokoban.release.R
import com.starsep.sokoban.release.Sokoban
import com.starsep.sokoban.release.controls.OnSwipeTouchListener
import com.starsep.sokoban.release.database.DatabaseManager
import com.starsep.sokoban.release.gamelogic.HighScore
import com.starsep.sokoban.release.gamelogic.Moves
import com.starsep.sokoban.release.model.GameModel
import kotlinx.android.synthetic.main.activity_game.*
import java.util.*

class GameActivity : SokobanActivity()
/*, GoogleApiClient.ConnectionCallbacks, GoogleApiClient.OnConnectionFailedListener*/ {
    // private lateinit var googleApiClient: GoogleApiClient
    private val gameModel: GameModel by lazy {
        ViewModelProviders.of(this).get(GameModel::class.java)
    }
    private val newGame: Boolean by lazy {
        intent.extras.getBoolean(Sokoban.NEW, false)
    }
    private val levelNumber: Int by lazy {
        intent.extras.getInt(Sokoban.LEVEL_NUMBER, 1)
    }

    private var timer: Timer? = null

    override fun onStart() {
        super.onStart()
        // googleApiClient.connect()
    }

    override fun onPause() {
        super.onPause()
        // googleApiClient.disconnect()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setupGameModel()
        setupUI()
        // googleApiClient = GoogleApiClientBuilder.build(this, gameView)
    }

    private fun setupUI() {
        setContentView(R.layout.activity_game)
        gameView.setOnTouchListener(object : OnSwipeTouchListener(baseContext) {
            override fun onSwipeRight() {
                gameModel.moveRight()
            }

            override fun onSwipeLeft() {
                gameModel.moveLeft()
            }

            override fun onSwipeTop() {
                gameModel.moveUp()
            }

            override fun onSwipeBottom() {
                gameModel.moveDown()
            }
        })

        resetButton.setOnClickListener {
            gameModel.resetLevel(it.context)
        }

        undoButton.setOnClickListener {
            gameModel.undoMove()
        }

        settingsButton.setOnClickListener {
            startActivity(Intent(this, SettingsActivity::class.java))
        }
    }

    private fun setupGameModel() {
        gameModel.startLevel(baseContext, levelNumber)
        gameModel.statsLive.observe(this, Observer<HighScore> { highScore ->
            val levelNumber = gameModel.levelNumber()
            val minutes = highScore!!.time / 60
            val seconds = highScore.time % 60
            val moves = highScore.moves
            val pushes = highScore.pushes
            statusTextView.text = String.format(getString(R.string.level_status),
                    levelNumber, minutes, seconds, moves, pushes)
        })
        gameModel.movesLive.observe(this, Observer<Moves> {
            DatabaseManager.instance(this).setCurrentGame(gameModel)
        })
        gameModel.wonLive.observe(this, Observer<Boolean> {
            timer?.let {
                it.cancel()
                timer = null
            }
        })
        gameModel.levelNumberLive.observe(this, Observer<Int> {
            timer = Timer()
            timer?.schedule(object : TimerTask() {
                override fun run() {
                    runOnUiThread {
                        gameModel.onSecondElapsed()
                    }
                }
            }, 0, 1000)
        })
    }

    fun isInEditMode(): Boolean {
        return gameView.isInEditMode
    }

    /*override fun onConnected(bundle: Bundle?) {
        Log.d(Sokoban.TAG, "onConnected(): connected to Google APIs")
        Games.setViewForPopups(googleApiClient, window.decorView.findViewById(android.R.id.content))
    }

    override fun onConnectionSuspended(i: Int) {

    }

    override fun onConnectionFailed(connectionResult: ConnectionResult) {
        Log.d(Sokoban.TAG, "onConnectionFailed(): attempting to resolve" + connectionResult)
        try {
            connectionResult.startResolutionForResult(this, 9143)
        } catch (e: IntentSender.SendIntentException) {
            e.printStackTrace()
        }
    }*/
}
